misaka/api.py

```diff
@@ -225,3 +259,112 @@ def __init__(self, nesting_level=6):
+    2. The URLs of links and images are filtered to prevent JavaScript injection.
+       See the :meth:`check_link` method below.
+    3. Optionally, the URLs can also be rewritten to counter other attacks such
+       as phishing.
+
+    Enabling URL rewriting requires extra arguments:
+
+    :arg link_rewrite: the URL of a redirect page, necessary to rewrite the
+        ``href`` attributes of links
+    :arg img_src_rewrite: the URL of an image proxy, necessary to rewrite the
+        ``src`` attributes of images
+
+    Both srings should include a ``{link}`` placeholder for the URL-encoded
+    target. Examples::
+
+        link_rewrite='https://example.com/redirect?url={link}',
+        img_src_rewrite='https://img-proxy-domain/{link}'
+    """
+    _allowed_url_re = re.compile(r'^https?:', re.I)
+
+    def __init__(self, flags=(), sanitization_mode='skip-html', nesting_level=0,
+                 link_rewrite=None, img_src_rewrite=None):
+        if not isinstance(flags, tuple):
+            raise TypeError("`flags` should be a tuple of strings")
+        HtmlRenderer.__init__(self, flags + (sanitization_mode,), nesting_level)
+        self.link_rewrite = link_rewrite
+        self.img_src_rewrite = img_src_rewrite
+
+    def autolink(self, raw_link, is_email):
+        """
+        Filters links generated by the ``autolink`` extension.
+        """
+        if self.check_link(raw_link):
+            link = self.rewrite_link(('mailto:' if is_email else '') + raw_link)
+            link = escape_html(link)
+            return '<a href="%s">%s</a>' % (link, escape_html(raw_link))
+        else:
+            return escape_html('<%s>' % raw_link)
+
+    def image(self, raw_link, title='', alt=''):
+        """
+        Filters the ``src`` attribute of an image.
+
+        Note that filtering the source URL of an ``<img>`` tag is only a very
+        basic protection, and it's mostly useless in modern browsers (they block
+        JavaScript in there by default). An example of attack that filtering
+        does not thwart is phishing based on HTTP Auth, see `this issue
+        <https://github.com/liberapay/liberapay.com/issues/504>`_ for details.
+
+        To mitigate this issue you should only allow images from trusted services,
+        for example your own image store, or a proxy (see :meth:`rewrite_link`).
+        """
+        if self.check_link(raw_link):
+            link = self.rewrite_link(raw_link, is_image_src=True)
+            maybe_alt = ' alt="%s"' % escape_html(alt) if alt else ''
+            maybe_title = ' title="%s"' % escape_html(title) if title else ''
+            link = escape_html(link)
+            return '<img src="%s"%s%s />' % (link, maybe_alt, maybe_title)
+        else:
+            return escape_html("![%s](%s)" % (alt, raw_link))
+
+    def link(self, content, raw_link, title=''):
+        """
+        Filters links.
+        """
+        if self.check_link(raw_link):
```

Comment:
When I enable the autolink extension email addresses are not rendered into links anymore.

My testing code:

```
from misaka import escape_html, Markdown, SaferHtmlRenderer, HtmlRenderer

render_normal = Markdown(HtmlRenderer(), extensions=('autolink',))
render = Markdown(SaferHtmlRenderer())
render_escape = Markdown(SaferHtmlRenderer(sanitization_mode='escape'))
renderer_rewrite = SaferHtmlRenderer(
    link_rewrite='//example.com/redirect/{link}',
    img_src_rewrite='//img_proxy/{link}',
)
render_rewrite = Markdown(renderer_rewrite, extensions=('autolink',),)
rewrite_link = renderer_rewrite.rewrite_link


print(render_normal('<https://b?x&y>'))
print(render_normal('Hello https://b?x&y Hola'))
print(render_normal('Hello https://example.com Hola'))
print(render_normal('Banana frank@61924.nl'))
print(render_normal('Banana <frank@61924.nl>'))

print('Safe:\n')

print(render_rewrite('<https://b?x&y>'))
print(render_rewrite('Hello https://b?x&y Hola'))
print(render_rewrite('Hello https://example.com Hola'))
print(render_rewrite('Banana frank@61924.nl'))
print(render_rewrite('Banana <frank@61924.nl>'))
```

Output:

```
<p><a href="https://b?x&amp;y">https://b?x&amp;y</a></p>

<p>Hello https://b?x&amp;y Hola</p>

<p>Hello <a href="https://example.com">https://example.com</a> Hola</p>

<p>Banana <a href="mailto:frank@61924.nl">frank@61924.nl</a></p>

<p>Banana <a href="mailto:frank@61924.nl">frank@61924.nl</a></p>

Safe:

<p><a href="//example.com/redirect/https%3A//b%3Fx%26y">https://b?x&amp;y</a></p>

<p>Hello https://b?x&amp;y Hola</p>

<p>Hello <a href="//example.com/redirect/https%3A//example.com">https://example.com</a> Hola</p>

<p>Banana &lt;frank@61924.nl&gt;</p>

<p>Banana &lt;frank@61924.nl&gt;</p>
```